proposals/shared-everything-threads/Overview.md

```diff
@@ -194,73 +91,373 @@ how a non-shared function accesses a `shared` memory):
+
+The `realm-local` option composes with the `async` option from JSPI. A shared function wrapper that is also async is async on every realm and thread.
+
+As an engine-internal optimization, setting a realm-local global to a realm-local function wrapper can skip a layer of indirection and set the global to refer directly to the bound function, since it will never change or be observed from another realm. This optimization only works if Wasm "thread-local" globals are actually realm-local in the Web embedding. True thread-local globals can be emulated on top of realm-local globals by ensuring that all realms on a thread trampoline through functions from a single realm to enter WebAssembly.
+
+[jspi]: https://github.com/WebAssembly/js-promise-integration
+[realm]: https://tc39.es/ecma262/#realm
+
+#### Waiter Queue API
+
+In the same way the linear memory wait and notify operations are mirrored and interoperable with those in JS, it will be useful to have JS methods to mirror the new `waitqueue.wait` and `waitqueue.notify` instructions. More importantly, it will also be useful to have an async version of the wait method analogous to `Atomics.waitAsync`.
+
+To expose this functionality, we add a new shareable JS type, `WebAssembly.WaitQueue`. Its realm-local prototype will have `wait`, `notify`, and `waitAsync` methods.
+
+> Note: If it turns out to be useful, we could also add methods for atomically accessing and modifying a wait queue's control field. In the current design, that can only be done from WebAssembly.
+
+### Binary format
+
+#### Memory orderings
+
+Just as the multi-memory proposal uses bit 6 of the memarg on memory access instructions to signify that a memory index follows, we will reserve bits 4-5 to encode the memory ordering (although we only use bit 4 in this proposal). For backward compatibility with the shared memory proposal, 0b00 will encode sequentially consistent ordering. 0b01 will encode release-acquire ordering. For memargs on non-atomic operations, the bits are not interpreted and must be 0b00. This scheme allows encoding alignments of up to 32768 bytes using bits 0-3, so there is no danger that we will need the newly reserved bits for alignment in the future, especially for atomic accesses.
+
+The new Instructions below do not have memarg immediates because they do not operate on memories. They instead have `u32:ordering` immediates, which are 0 for sequentially-consistent ordering or 1 for release-acquire ordering. `atomic.fence` already has a reserved zero byte immediate, which we now interpret as a u32:ordering immediate, allowing us to express release-acquire fences as well.
```

Comment:
```suggestion
The new instructions below do not have `memarg` immediates because they do not operate on memories. They instead have `u32:ordering` immediates, which are 0 for sequentially-consistent ordering or 1 for release-acquire ordering. `atomic.fence` already has a reserved zero byte immediate, which we now interpret as a u32:ordering immediate, allowing us to express release-acquire fences as well.
```